scheduler/src/cook/mesos/constraints.clj

```diff
@@ -155,7 +157,41 @@
   [job]
   (->user-defined-constraint (:job/constraint job)))
 
-(def job-constraint-constructors [build-novel-host-constraint build-gpu-host-constraint build-user-defined-constraint])
+(defrecord estimated-completion-constraint [estimated-end-time host-lifetime-mins]
+  JobConstraint
+  (job-constraint-name [this] (get-class-name this))
+  (job-constraint-evaluate
+    [this _ vm-attributes _]
+    (job-constraint-evaluate this _ vm-attributes))
+  (job-constraint-evaluate
+    [this _ vm-attributes]
+    (if-let [host-start-time (get vm-attributes "host-start-time")]
+      (let [host-death-time (+ (* 1000 (long host-start-time))
+                               (* 60 1000 host-lifetime-mins))]
+        (if (< host-death-time estimated-end-time)
+          [true ""]
+          [false "host is expected to shutdown before job completion"]))
+      [true ""])))
+
+(defn build-estimated-completion-constraint
```

Comment:
Seems like the logic in here is complex enough to warrant a docstring here

Reply:
Done